pkg/trace/writer/newwriter/testserver.go

```diff
@@ -77,14 +79,32 @@ func (rs *requestStatus) nextResponse() int {
 	return statusCode
 }
 
-func (ts *testServer) Failed() int   { return int(atomic.LoadUint64(&ts.failed)) }
-func (ts *testServer) Retried() int  { return int(atomic.LoadUint64(&ts.retried)) }
-func (ts *testServer) Total() int    { return int(atomic.LoadUint64(&ts.total)) }
+// Peak returns the maximum number of simultaneous connections that were active
+// while the server was running.
+func (ts *testServer) Peak() int { return int(atomic.LoadInt64(&ts.peak)) }
+
+// Failed returns the number of connections to which the server responded with an
+// HTTP status code that is non-2xx and non-5xx.
+func (ts *testServer) Failed() int { return int(atomic.LoadUint64(&ts.failed)) }
+
+// Failed returns the number of connections to which the server responded with a
+// 5xx HTTP status code.
+func (ts *testServer) Retried() int { return int(atomic.LoadUint64(&ts.retried)) }
+
+// Total returns the total number of connections which reached the server.
+func (ts *testServer) Total() int { return int(atomic.LoadUint64(&ts.total)) }
+
+// Failed returns the number of connections to which the server responded with a
+// 2xx HTTP status code.
 func (ts *testServer) Accepted() int { return int(atomic.LoadUint64(&ts.accepted)) }
 
 // ServeHTTP responds based on the request body.
 func (ts *testServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	atomic.AddUint64(&ts.total, 1)
+	if v := atomic.AddInt64(&ts.active, 1); v > atomic.LoadInt64(&ts.peak) {
+		atomic.SwapInt64(&ts.peak, v)
```

Comment:
Peak is not guaranteed to be correct with this:
* Routine 1 comes in bumps active to 1, loads peak (0) and is scheduled out.
* Routine 2 comes in bumps active to 2, loads peak (0) and is scheduled out.
* Routine 3 comes in bumps active to 3, loads peak (0) and sets peak to 3.
* Routine 2 comes back in and sets peak to 2.
* Routine 1 comes back in and sets peak to 1.

You need an exclusivity zone surrounding the reading and the writing for this to be correct. Since this is a test server only used for testing I don't think it'd be much of a problem if you added a lock here.

Reply:
I'm not a huge expert on the scheduler, but I remember reading somewhere that the scheduler jumps to another routine only when encountering a blocking operation. Nevertheless, I tried what you said and it didn't work much better. I discovered however that it was because the requests were simply server too fast, so I've added an option to create some latency. After doing that, all calls of that test reached the maximum set limit (100). So I think that should do it.

Pushed a new commit with that.

Reply:
I don't usually like to write code based on assumptions of how some piece of the runtime behaves at the current time. Plus, if what you said were true then a routine running an infinite loop would prevent all other routines from running. 

Anyway, I don't think it's much of a matter for this testing code but if this were on production code I'd really prefer the theoretically correct version.

Reply:
I 100% agree with that. We shouldn't rely on how the scheduler is assumed to work when writing code, and it should be correct. If you want me to add the lock, happy to do so. I only removed it because it seemed to work without it too, after adding fake latency.

(Side note: I do remember reading some interesting things around `for{}` vs `select{}`, something along the lines: https://stackoverflow.com/questions/18661602/what-does-an-empty-select-do).